### ISSUE TITLE
Internal Error for "Convert to enhanced 'for' loops" clean-up without "Only if loop variable used"

### ISSUE THREAD

**Issue by HannesWell** [eclipse-jdt/eclipse.jdt.ui#120]
I tried to apply the "Convert to enhanced 'for' loops" clean-up without "Only if loop variable used" being selected to the equinox code base and it failed with the error below. If the second option is selected everything works fine.

Furthermore I noticed that the list of selected actions in the Use custom profile section (reached via Right click on project -> Source -> Clean up...) lists the action twice:
```
- Convert to enhanced 'for' loops
- Convert to enhanced 'for' loops
```
If I additionally select "Only if loop variable used" the list looks like this (which actually also seems duplicated):
```
Convert to enhanced 'for' loops only if the loop variable is used
Convert to enhanced 'for' loops
```
Product:
```
Eclipse SDK

Version: 2022-06 (4.25)
Build id: I20220624-1800
```
Exception Stack Trace of "Internal Error":
```
java.lang.reflect.InvocationTargetException
	at org.eclipse.jface.operation.ModalContext.run(ModalContext.java:395)
	at org.eclipse.jface.wizard.WizardDialog.run(WizardDialog.java:1033)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizard.createChange(RefactoringWizard.java:662)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizard.computeUserInputSuccessorPage(RefactoringWizard.java:504)
	at org.eclipse.ltk.ui.refactoring.UserInputWizardPage.computeSuccessorPage(UserInputWizardPage.java:77)
	at org.eclipse.ltk.ui.refactoring.UserInputWizardPage.getNextPage(UserInputWizardPage.java:113)
	at org.eclipse.jdt.internal.ui.fix.CleanUpRefactoringWizard$CleanUpConfigurationPage.getNextPage(CleanUpRefactoringWizard.java:437)
	at org.eclipse.jface.wizard.WizardDialog.nextPressed(WizardDialog.java:910)
	at org.eclipse.jface.wizard.WizardDialog.buttonPressed(WizardDialog.java:468)
	at org.eclipse.jface.dialogs.Dialog.lambda$0(Dialog.java:619)
	at org.eclipse.swt.events.SelectionListener$1.widgetSelected(SelectionListener.java:84)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:252)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4251)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1066)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4068)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3645)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:823)
	at org.eclipse.jface.window.Window.open(Window.java:799)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizardOpenOperation.lambda$0(RefactoringWizardOpenOperation.java:190)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:74)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizardOpenOperation.run(RefactoringWizardOpenOperation.java:209)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizardOpenOperation.run(RefactoringWizardOpenOperation.java:126)
	at org.eclipse.jdt.internal.ui.refactoring.actions.RefactoringStarter.activate(RefactoringStarter.java:41)
	at org.eclipse.jdt.internal.corext.refactoring.RefactoringExecutionStarter.startCleanupRefactoring(RefactoringExecutionStarter.java:257)
	at org.eclipse.jdt.internal.ui.actions.AllCleanUpsAction.performRefactoring(AllCleanUpsAction.java:71)
	at org.eclipse.jdt.internal.ui.actions.CleanUpAction.runOnMultiple(CleanUpAction.java:215)
	at org.eclipse.jdt.internal.ui.actions.CleanUpAction.run(CleanUpAction.java:119)
	at org.eclipse.jdt.ui.actions.SelectionDispatchAction.dispatchRun(SelectionDispatchAction.java:274)
	at org.eclipse.jdt.ui.actions.SelectionDispatchAction.run(SelectionDispatchAction.java:252)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:414)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4251)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1066)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4068)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3645)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
Caused by: java.lang.IllegalArgumentException: Invalid identifier : >List<V><
	at org.eclipse.jdt.core.dom.SimpleName.setIdentifier(SimpleName.java:247)
	at org.eclipse.jdt.core.dom.AST.newSimpleName(AST.java:2777)
	at org.eclipse.jdt.core.dom.AST.newName(AST.java:2427)
	at org.eclipse.jdt.internal.corext.fix.helper.AbstractTool.addImport(AbstractTool.java:66)
	at org.eclipse.jdt.internal.corext.fix.helper.WhileToForEach.rewrite(WhileToForEach.java:424)
	at org.eclipse.jdt.internal.corext.fix.helper.WhileToForEach.rewrite(WhileToForEach.java:1)
	at org.eclipse.jdt.internal.corext.fix.UseIteratorToForLoopFixCore$1.rewriteAST(UseIteratorToForLoopFixCore.java:77)
	at org.eclipse.jdt.internal.corext.fix.CompilationUnitRewriteOperationsFixCore.createChange(CompilationUnitRewriteOperationsFixCore.java:98)
	at org.eclipse.jdt.internal.ui.fix.CleanUpFixWrapper.createChange(CleanUpFixWrapper.java:46)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.calculateChange(CleanUpRefactoring.java:775)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpASTRequestor.calculateSolutions(CleanUpRefactoring.java:301)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpASTRequestor.acceptAST(CleanUpRefactoring.java:279)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.resolve(CompilationUnitResolver.java:1008)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.resolve(CompilationUnitResolver.java:692)
	at org.eclipse.jdt.core.dom.ASTParser.createASTs(ASTParser.java:954)
	at org.eclipse.jdt.internal.corext.dom.ASTBatchParser.createASTs(ASTBatchParser.java:82)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpFixpointIterator.next(CleanUpRefactoring.java:399)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.cleanUpProject(CleanUpRefactoring.java:682)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.checkFinalConditions(CleanUpRefactoring.java:642)
	at org.eclipse.ltk.core.refactoring.CheckConditionsOperation.run(CheckConditionsOperation.java:86)
	at org.eclipse.ltk.core.refactoring.CreateChangeOperation.run(CreateChangeOperation.java:122)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2380)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2405)
	at org.eclipse.ltk.internal.ui.refactoring.WorkbenchRunnableAdapter.run(WorkbenchRunnableAdapter.java:89)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:122)
Root exception:
java.lang.IllegalArgumentException: Invalid identifier : >List<V><
	at org.eclipse.jdt.core.dom.SimpleName.setIdentifier(SimpleName.java:247)
	at org.eclipse.jdt.core.dom.AST.newSimpleName(AST.java:2777)
	at org.eclipse.jdt.core.dom.AST.newName(AST.java:2427)
	at org.eclipse.jdt.internal.corext.fix.helper.AbstractTool.addImport(AbstractTool.java:66)
	at org.eclipse.jdt.internal.corext.fix.helper.WhileToForEach.rewrite(WhileToForEach.java:424)
	at org.eclipse.jdt.internal.corext.fix.helper.WhileToForEach.rewrite(WhileToForEach.java:1)
	at org.eclipse.jdt.internal.corext.fix.UseIteratorToForLoopFixCore$1.rewriteAST(UseIteratorToForLoopFixCore.java:77)
	at org.eclipse.jdt.internal.corext.fix.CompilationUnitRewriteOperationsFixCore.createChange(CompilationUnitRewriteOperationsFixCore.java:98)
	at org.eclipse.jdt.internal.ui.fix.CleanUpFixWrapper.createChange(CleanUpFixWrapper.java:46)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.calculateChange(CleanUpRefactoring.java:775)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpASTRequestor.calculateSolutions(CleanUpRefactoring.java:301)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpASTRequestor.acceptAST(CleanUpRefactoring.java:279)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.resolve(CompilationUnitResolver.java:1008)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.resolve(CompilationUnitResolver.java:692)
	at org.eclipse.jdt.core.dom.ASTParser.createASTs(ASTParser.java:954)
	at org.eclipse.jdt.internal.corext.dom.ASTBatchParser.createASTs(ASTBatchParser.java:82)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpFixpointIterator.next(CleanUpRefactoring.java:399)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.cleanUpProject(CleanUpRefactoring.java:682)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.checkFinalConditions(CleanUpRefactoring.java:642)
	at org.eclipse.ltk.core.refactoring.CheckConditionsOperation.run(CheckConditionsOperation.java:86)
	at org.eclipse.ltk.core.refactoring.CreateChangeOperation.run(CreateChangeOperation.java:122)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2380)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2405)
	at org.eclipse.ltk.internal.ui.refactoring.WorkbenchRunnableAdapter.run(WorkbenchRunnableAdapter.java:89)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:122)
```

**Comment by HannesWell** [comment]
OK, I was able to boil it down to the following method:
```
private static <K, V> List<V> method(Map<K, List<V>> map) {
	List<V> results = new ArrayList<>();
	Iterator<List<V>> iterator = map.values().iterator();
	while (iterator.hasNext()) {
		results.addAll(iterator.next());
	}
	return results;
}
```
When running only `Convert to enhanced 'for' loops` without the option `only if the loop variable is used` then the clean up fails with the error above (even in the preview).

Furthermore if I select that option and run `Convert to enhanced 'for' loops only if the loop variable is used` that method is not changed, but I would expect that it is converted to something like:
```
private static <K, V> List<V> expected(Map<K, List<V>> map) {
	List<V> results = new ArrayList<>();
	for (List<V> value : map.values()) {
		results.addAll(value);
	}
	return results;
}
```

**Comment by carstenartur** [comment]
Thanks again for your patience and continued testing!

Just a short remark: That there are two identical entries in the list is because there are two different cleanups bound to the identical constant

```
import static org.eclipse.jdt.internal.corext.fix.CleanUpConstants.CONTROL_STATEMENTS_CONVERT_FOR_LOOP_ONLY_IF_LOOP_VAR_USED;
import static org.eclipse.jdt.internal.corext.fix.CleanUpConstants.CONTROL_STATEMENTS_CONVERT_FOR_LOOP_TO_ENHANCED;

```
![image](https://user-images.githubusercontent.com/3164220/175777094-2cc857ee-9946-4830-97dc-e1d5a0ae2972.png)

For that the two different text messages do not need to be equal and the have not been in earlier versions. But now both make use of the same text

![image](https://user-images.githubusercontent.com/3164220/175777201-2830b2b5-5066-4638-9e0d-b8f738fca795.png)


https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/55/files#diff-fe82b809bb16e49eedbb27bc774782b47cd4c9c44e9bce17087829f21d36fdf6

If it is up to me we can discuss if there is a better way but lets hear what @jjohnstn thinks about it.

**Comment by HannesWell** [comment]
> Just a short remark: That there are two identical entries in the list is because there are two different cleanups bound to the identical constant
> 
So the two cleanup actions use the same constant and are (de-)activated by the same checkbox in the UI, aren't they?
Then I would consider it an implementation detail that there are actually two actions and present just one item in the UI.


First of all @ktatavarthi, thank you for your fix. I just opened a PR to add the presented reproducer as a JUnit test case. This test case also shows that there is a false import added with that fix. I was also able to reproduce that in an Editor of an Eclipse-app launched from my JDT workspace.

> > Furthermore if I select that option and run `Convert to enhanced 'for' loops only if the loop variable is used` that method is not changed
> 
> Regarding the above statement, As shown in the `cleanup preview of Convert to enhanced 'for' loops only if the loop variable is used`, we see that there is no loop variable.

Yes after thinking about that I also came to the conclusion that I understood it wrong.
I falsely assumed that this was referring to loops that use a counter variable like in the following snippet or similar for an array:
```
	List<String> list = List.of("");
	for (int i = 0; i < list .size(); i++) {
		System.out.println("Hello World");
	}
```
In these cases it would IMHO make sense to not use an enhanced for loop to avoid the iterator creation and I assumed that this was the motivation for the option `only if the loop variable is used`. In cases where the result of Iterator.next() is passed directly to a method I think it still makes sense to convert to an enhanced for-loop since the iterator is used anyway.
If the motivation is really to avoid iterator creation it might be reasonable to refine the criteria for this sub-option to only skip the conversion to an enhanced loop if the loop variable **after** applying the clean up would not be used.
